rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/WorkerBookKeeper.java

```diff
@@ -14,28 +14,80 @@
 package com.qubole.rubix.bookkeeper;
 
 import com.codahale.metrics.MetricRegistry;
+import com.google.common.base.Ticker;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.qubole.rubix.core.ClusterManagerInitilizationException;
+import com.qubole.rubix.core.utils.ClusterUtil;
 import com.qubole.rubix.spi.BookKeeperFactory;
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.ClusterType;
+import com.qubole.rubix.spi.RetryingBookkeeperClient;
 import com.qubole.rubix.spi.thrift.HeartbeatStatus;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.conf.Configuration;
 
 import java.io.FileNotFoundException;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 public class WorkerBookKeeper extends BookKeeper
 {
   private static Log log = LogFactory.getLog(WorkerBookKeeper.class);
+  private LoadingCache<Integer, List<String>> nodeListCache;
   private HeartbeatService heartbeatService;
+  private BookKeeperFactory bookKeeperFactory;
+  private RetryingBookkeeperClient client;
+  // The hostname of the master node.
+  private String masterHostname;
 
   public WorkerBookKeeper(Configuration conf, MetricRegistry metrics) throws FileNotFoundException
   {
-    this(conf, metrics, new BookKeeperFactory());
+    this(conf, metrics, Ticker.systemTicker(), new BookKeeperFactory());
   }
 
-  public WorkerBookKeeper(Configuration conf, MetricRegistry metrics, BookKeeperFactory factory) throws FileNotFoundException
+  public WorkerBookKeeper(Configuration conf, MetricRegistry metrics, Ticker ticker, BookKeeperFactory factory) throws FileNotFoundException
   {
-    super(conf, metrics);
+    super(conf, metrics, Ticker.systemTicker());
+    this.bookKeeperFactory = factory;
+    this.masterHostname = ClusterUtil.getMasterHostname(conf);
     startHeartbeatService(conf, metrics, factory);
+    initializeNodesCache(conf, ticker);
+  }
+
+  private void initializeNodesCache(final Configuration conf, final Ticker ticker)
+  {
+    int expiryPeriod = CacheConfig.getWorkerNodeInfoExpiryPeriod(conf);
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    nodeListCache = CacheBuilder.newBuilder()
+        .ticker(ticker)
+        .expireAfterWrite(expiryPeriod, TimeUnit.SECONDS)
```

Comment:
refreshAfterWrite would suit better than expireAfterWrite. Node additions and removals are less frequent and we would pay a penalty of reloading the value every 5min, async refresh with a fallback to direct path once in a while would be okay

Reply:
+1

Reply:
Done.